File: cses/04_graph_algorithms/1196_flight-routes/solution.py
import sys
from heapq import heappush, heappop


def main():
    # Đọc toàn bộ input theo buffer cho nhanh.
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    m = int(data[idx]); idx += 1
    k = int(data[idx]); idx += 1

    # Đồ thị có hướng dạng adjacency list: mỗi cạnh là (v, w) = (đỉnh đích, giá vé).
    adj = [[] for _ in range(n + 1)]
    for _ in range(m):
        a = int(data[idx]); b = int(data[idx + 1]); c = int(data[idx + 2])
        idx += 3
        adj[a].append((b, c))

    # cnt[u] = số lần đỉnh u đã được pop và finalize (tối đa k lần).
    cnt = [0] * (n + 1)
    heap = [(0, 1)]              # min-heap gồm các trạng thái (khoảng cách, đỉnh)
    results = []

    # Gắn hàm hot-loop vào biến cục bộ để giảm chi phí tra cứu mỗi vòng lặp.
    push = heappush
    pop = heappop

    # Dijkstra mở rộng: dừng khi đã thu đủ k đường đi rẻ nhất tới đỉnh n.
    while heap and len(results) < k:
        d, u = pop(heap)
        # Đỉnh này đã đủ k khoảng cách tốt nhất, bỏ qua.
        if cnt[u] >= k:
            continue
        # Lần pop hợp lệ thứ i của u cho đường đi ngắn thứ i tới u.
        cnt[u] += 1
        if u == n:
            results.append(d)
            if len(results) == k:
                break
        # Mở rộng sang các đỉnh kề chưa đủ k lần finalize.
        for v, w in adj[u]:
            if cnt[v] < k:
                push(heap, (d + w, v))

    # In k tổng giá theo thứ tự không giảm.
    sys.stdout.write(' '.join(map(str, results)))


main()

File: cses/04_graph_algorithms/1196_flight-routes/solution.cpp
#include <functional>
#include <iostream>
#include <queue>
#include <utility>
#include <vector>

struct Edge {
    int destination;
    long long cost;
};

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, m, k;
    std::cin >> n >> m >> k;

    // Đồ thị có hướng dạng adjacency list: cạnh a -> b với giá vé cost.
    std::vector<std::vector<Edge>> graph(n + 1);
    for (int i = 0; i < m; ++i) {
        int a, b;
        long long cost;
        std::cin >> a >> b >> cost;
        graph[a].push_back({b, cost});
    }

    // Dijkstra mở rộng cho K shortest paths.
    // State = (khoảng cách, đỉnh); min-heap luôn lấy khoảng cách nhỏ nhất trước.
    using State = std::pair<long long, int>;
    std::priority_queue<State, std::vector<State>, std::greater<State>> queue;
    // finalized_count[u] = số lần đỉnh u đã được pop và finalize (tối đa k lần).
    std::vector<int> finalized_count(n + 1, 0);
    std::vector<long long> answers;
    answers.reserve(k);
    queue.emplace(0, 1);

    // Dừng khi đã thu đủ k đường đi rẻ nhất tới đỉnh n.
    while (!queue.empty() && static_cast<int>(answers.size()) < k) {
        const auto [distance, node] = queue.top();
        queue.pop();
        // Đỉnh này đã có đủ k khoảng cách tốt nhất, không cần xử lý thêm.
        if (finalized_count[node] >= k) {
            continue;
        }

        // Lần pop hợp lệ thứ i của node cho đường đi ngắn thứ i tới node.
        ++finalized_count[node];
        if (node == n) {
            answers.push_back(distance);
        }

        // Mở rộng sang các đỉnh kề chưa đủ k lần finalize.
        for (const Edge& edge : graph[node]) {
            if (finalized_count[edge.destination] < k) {
                queue.emplace(distance + edge.cost, edge.destination);
            }
        }
    }

    // In k tổng giá theo thứ tự không giảm (heap đã bảo đảm thứ tự này).
    for (int i = 0; i < static_cast<int>(answers.size()); ++i) {
        if (i > 0) {
            std::cout << ' ';
        }
        std::cout << answers[i];
    }
    std::cout << '\n';
    return 0;
}
